mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+
+### Allow mapping an `MLTensor` to script
+
+Why doesn't `MLTensor` have a [`mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync) method, as a `GPUBuffer` does?
+
+`MLTensor`s are used as inputs and outputs to machine learning models. In practice, we expect the size of the model inputs and outputs to be dwarfed by the size of the model weights, which are uploaded via the `MLGraphBuilder.constant()` method. We may re-evaluate this stance in the future if we discover that reading and writing data to an `MLTensor` is a bottleneck, though even in that case we may prefer to explore a solution which bypasses `ArrayBuffer`s altogether, as discussed [above](#support-passing-an-mltensor-as-a-constant).
+
+### Hash input buffer contents
+
+One approach to solve the [buffer reuse](#buffer-reuse) case is for the WebNN implementation to silently avoid making redundant data copies if the same buffer contents are repeatedly passed as inputs. This may be achieved by hashing the contents of each input. This approach has downsides, such as [managing the extra copies in the hash map](#resource-management) and that hashing the buffer contents may be expensive as it requires reading the entire input. This approach also does not address the other use cases.
+
+## References & Acknowledgements
+
+Many thanks for valuable feedback and advice from:
+
+- Bryan Bernhart
+- Joshua Bell
+- Mike Wyrzykowski
+- Ningxin Hu
+- Phillis Tang
+- Rafael Cintron
+- Reilly Grant
+- Zoltan Kis
+
+---
+
+## Appendix
+
+### Tentative IDL
+
+```javascript
+typedef [EnforceRange] unsigned long MLTensorUsageFlags;
+
+namespace MLTensorUsage {
+    const MLFlagsConstant READ_FROM      = 0x0001;
+    const MLFlagsConstant WRITE_TO       = 0x0002;
+    const MLFlagsConstant WEBGPU_INTEROP = 0x0004;
+};
+
+dictionary MLTensorDescriptor : MLOperandDescriptor {
+  required MLTensorUsageFlags usage;
+};
+
+typedef record<DOMString, MLTensor> MLNamedTensors;
+
+interface MLTensor {
+  readonly attribute MLOperandDataType dataType;
+  readonly attribute FrozenArray<unsigned long> shape;
+  readonly attribute unsigned long MLTensorUsageFlags usage;
+
+  void destroy();
+};
+
+partial interface MLContext {
+  Promise<MLTensor> createTensor(MLTensorDescriptor descriptor);
+
+  void writeBuffer(MLTensor dstTensor, [AllowShared] ArrayBuffer srcData);
+  void writeBuffer(MLTensor dstTensor, [AllowShared] ArrayBufferView srcData);
+  
+  Promise<ArrayBuffer> readBuffer(MLTensor srcTensor);
+  Promise<void> readBuffer(MLTensor srcTensor, [AllowShared] ArrayBuffer dstData);
+  Promise<void> readBuffer(MLTensor srcTensor, [AllowShared] ArrayBufferView dstData);
+  
+  void dispatch(MLGraph graph, MLNamedTensors inputs, MLNamedTensors outputs);
+};
+
+// For WebGPU Interop
+
+interface GPUExternalBuffer {};
```

Comment:
What’s the benefit of having a new `GPUExternalBuffer` object? It doesn’t seem like `GPUExternalBuffer` represents something significantly different from a normal `GPUBuffer`. I think we should consider if a new type is really necessary - because if we could get away without a new type and just overload/restrict `GPUBuffer` it would save us the WebGPU and WGSL implementation work.

Reply:
@Kangz do you have any opinions on this?